components/object_store/src/multipart.rs

```diff
@@ -0,0 +1,258 @@
+        }
+    }
+}
+
+/// The process of ObjectStore write multipart upload is:
+/// First, Call ObjectStore::multi_upload to begin multipart upload.
+/// Second, Call AsyncWrite::poll_write to wrtie buffer data to oject,this
+/// fuction can be called many times. Last, Call AsyncWrite::poll_shutdown to
+/// finish current mulipart upload. See more in
+/// [`analytic_engine::sst::parquet::writer::ParquetSstWriter::write`]
+impl<T> CloudMultiPartUpload<T>
+where
+    T: CloudMultiPartUploadImpl + Send + Sync,
+{
+    // The `poll_flush` function will only flush the in-progress tasks.
+    // The `final_flush` method called during `poll_shutdown` will flush
+    // the `current_buffer` along with in-progress tasks.
+    fn final_flush(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Result<(), io::Error>> {
+        // Poll current tasks
+        self.as_mut().poll_tasks(cx)?;
+
+        // If current_buffer is not empty, see if it can be submitted
+        if !self.current_buffer.is_empty() && self.tasks.len() < self.max_concurrency {
+            self.as_mut().final_flush_buffer();
+        }
+
+        self.as_mut().poll_tasks(cx)?;
+
+        // If tasks and current_buffer are empty, return Ready
+        if self.tasks.is_empty() && self.current_buffer.is_empty() {
+            Poll::Ready(Ok(()))
+        } else {
+            Poll::Pending
+        }
+    }
+}
+
+impl<T> AsyncWrite for CloudMultiPartUpload<T>
+where
+    T: CloudMultiPartUploadImpl + Send + Sync,
+{
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        // Poll current tasks
+        self.as_mut().poll_tasks(cx)?;
+
+        // If adding buf to pending buffer would trigger send, check
+        // whether we have capacity for another task.
+        let part_size = self.part_size;
+        let enough_to_send = (buf.len() + self.current_buffer.len()) >= part_size;
+        if enough_to_send && self.tasks.len() < self.max_concurrency {
+            // If we do, copy into the buffer and submit the task, and return ready.
+            self.current_buffer.extend_from_slice(buf);
+            // Flush buffer data, use custom method
+            self.as_mut().flush_buffer();
+            // We need to poll immediately after adding to setup waker
+            self.as_mut().poll_tasks(cx)?;
+
+            Poll::Ready(Ok(buf.len()))
+        } else if !enough_to_send {
+            self.current_buffer.extend_from_slice(buf);
+            Poll::Ready(Ok(buf.len()))
+        } else {
+            // Waker registered by call to poll_tasks at beginning
+            Poll::Pending
+        }
+    }
+
+    fn poll_flush(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Result<(), io::Error>> {
+        // Poll current tasks
+        self.as_mut().poll_tasks(cx)?;
```

Comment:
Need to call `final_flush` here? If not, how to ensure the remaining bytes in the `current_buffer` flushed?